aten/src/ATen/native/README.md

```diff
@@ -0,0 +1,109 @@
+The C++ function declarations won't match the declaration here because
+they will undergo the standard ATen C++ transformations, e.g. use of const-ref
+for non-inplace Tensor arguments (instead of `Tensor` you will
+take `const Tensor&`).  So after you finish writing your header, we recommend
+running a build and getting the declaration from `NativeFunctions.h`
+(find it with `find -name NativeFunctions.h`) to copy paste into your
+C++ file.
+
+The declarations also support the following attributes:
+
+```
+variants: function, method
+```
+
+Controls whether Tensor method (`t.foo()`) or namespace Function (`at::foo()`) is
+generated as a result of this declaration.  If the declaration is a method,
+you must have an argument `Tensor self` as the first argument.  In general, you
+should default to defining a new ATen function as `variants: function`, unless
+you know you want something to be usable as a method.
+
+```
+dispatch:
+    CPU: func_cpu
+    CUDA: func_cuda
+```
+
+This specifies the actual name of the function you want to dispatch to, so you
+can dispatch to different functions depending on whether or not you have CPU or
+CUDA tensors.  Technically, it is also possible to write `dispatch: func_name`
+to unconditionally dispatch to a native function whose name is different than
+the name in the public ATen API, but this is generally frowned upon (just name
+them the same thing!)
+
+```
+python_default_init:
+  argument_name: initializing_expression
+```
+
+A map from argument names to default initialize expressions in C++. Such default
+expressions will only be used in Python API. This allows us to write argument
+with a default value that can either cause ambiguity in C++ (e.g., `Scalar p`
+argument in `norm`) or have a type that doesn't allow default value
+None/NULL/nullptr (e.g., `int64_t fft_size` argument in stft, which we want to
+default to value of another argument if not provided).
+
+## Writing the implementation
+
+Implementations of native functions go in an appropriate C++ file in the
+`native/` directory (they are organized roughly by topic, but there is no
+semantic meaning to their organization, except for CUDA files, which must
+go in `cuda`.)  To write a native function, you only need to write a C++
+implementation (no header necessary) with a matching signature to
+the generated header from the ATen metadata.  There are many
+simple native functions; take a look at some of them to see what to do.
+
+There are some important gotchas which are important to keep in mind when
+writing a native function:
+
+* NEVER EVER EVER use `at::CPU` or `at::CUDA` directly; instead, create tensors
+  from the `type()` of one of the input tensors, e.g., `input.type().tensor()`
+  or `input.type().toScalarType(kByte)` if you need a tensor type directly.
+  See https://github.com/pytorch/pytorch/issues/4477 for more details.
+
+* Keep in mind whether or not your function must have an explicit derivative
```

Comment:
If I was just starting out, I wouldn't know how to analyze this.  How do I know if my function is automatically differentiable or not?  (Maybe suggest they write an autograd test for it?)